### PR TITLE
Feature/parallelisation

### PR DESCRIPTION
**Pull request by grayseff**
Extensive changes made to implement parallelised code:

 This includes several statistical tools, changes to indices etc. Individual comments will be assigned.

### CODE REVIEW

File: .vscode/settings.json
Comment (grayseff) [inline .vscode/settings.json:2]: This change automatically primes all threads rather than just one.

File: scripts/run_ne13.jl
Comment (grayseff) [inline scripts/run_ne13.jl:18]: change to cycles

File: scripts/run_ne13.jl
Comment (grayseff) [inline scripts/run_ne13.jl:81]: includes timing function at the moment

File: src/EnergyEvaluation.jl
Comment (grayseff) [inline src/EnergyEvaluation.jl:11]: Required to run RuNNer

File: src/EnergyEvaluation.jl
Comment (grayseff) [inline src/EnergyEvaluation.jl:13]: Names the new exported potentials

File: src/EnergyEvaluation.jl
Comment (grayseff) [inline src/EnergyEvaluation.jl:54]: These types relate to individual ML potentials.

File: src/EnergyEvaluation.jl
Comment (grayseff) [inline src/EnergyEvaluation.jl:98]: renamed delta_en for clarity later

File: src/EnergyEvaluation.jl
Comment (grayseff) [inline src/EnergyEvaluation.jl:102]: function for independent RuNNer energy updates

File: src/InputParams.jl
Comment (grayseff) [inline src/InputParams.jl:104]: Functional RDF vector now

File: src/MCMoves.jl
Comment (grayseff) [inline src/MCMoves.jl:6]: update_max_stepsize as defined here is not the one used in MC_run, this isn't a change I made, it's just defined twice.

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:15]: Required to use ML potentials

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:9]: new dependencies required for saving

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:5]: exporting more functions for redefinition in parallelrun

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:68]: split MCState definition into individual potentials

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:103]: renamed delta_en to avoid global variable names

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:129]: more redefinition of variable names

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:153]: This function now has a minimum and maximum acceptance rate we can define ourselves

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:229]: more redefining variable names

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:253]: redefinition to avoid using the name ran

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:274]: removed @inbounds

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:302]: this version of mc_cycle works with RuNNer

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:353]: Moved sampling from MC_Run to its own function

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:445]: The functions ABOVE this line are the save files used to reinitialise MC_Run

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:450]: Two methods for initialising histograms

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:572]: restart not yet complete

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:576]: sets the energy and E^2 counters

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:600]: absurdly high and low values are used for initialising the min/max energies. This is required for initialising the histograms

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:608]: this section determines whether we update e_min e_max throughout equilibration

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:634]: initialises the histograms

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:641]: checkpoint created at the end of equilibration

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:652]: cycles are now governed by the PTMC_cycle function, with save_ham determining whether we progressively store energy values or whether we use a running total/histograms

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:662]: index i is important for save functions typically, so we start from the restart_index

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:676]: This part relates to post-statistics on the entire vector of energy values

File: src/MCRun.jl
Comment (grayseff) [inline src/MCRun.jl:679]: This part relates to post-statistics on the en_total stored throughout the run

File: src/ParallelTemperingMonteCarlo.jl
Comment (grayseff) [inline src/ParallelTemperingMonteCarlo.jl:4]: self explanatory introduction of new dependencies

File: src/multihist.jl
Comment (grayseff) [inline src/multihist.jl:59]: This determines whether we have "spillover" bins

File: src/multihist.jl
Comment (grayseff) [inline src/multihist.jl:268]: No longer begins at T=0, begins at the lowest T of the simulation.

File: src/multihist.jl
Comment (grayseff) [inline src/multihist.jl:378]: Included slashes to avoid problems with pwd()

File: Project.toml
Comment (grayseff) [inline Project.toml:12]: removed plots from project.toml

File: src/RuNNer.jl
Comment (ElkePahl) [inline src/RuNNer.jl:130]: why is atomtype once a vector and once a string?

File: src/RuNNer.jl
Comment (ElkePahl) [inline src/RuNNer.jl:170]: what is the usage of third method for writefile, what is 'ix'
I am a bit at a loss here: testconfig seems to be a position only (not a Config) but then it become an SVector?

File: src/RuNNer.jl
Comment (ElkePahl) [inline src/RuNNer.jl:174]: again be more specific when naming and add a docu string

File: src/MCRun.jl
Comment (ElkePahl) [reply src/MCRun.jl:5]: ?

File: src/MCRun.jl
Comment (ElkePahl) [inline src/MCRun.jl:77]: what happened to the method for DFTPotential?

File: src/MCRun.jl
Comment (ElkePahl) [reply src/MCRun.jl:105]: Again: I like delta_en or delta_energy more as code easier to read, and I don't think that it is an issue
I am also happy to have it more specific to distinguish between different kinds of delta_en if needed

File: src/MCRun.jl
Comment (ElkePahl) [reply src/MCRun.jl:155]: that's good

File: src/MCRun.jl
Comment (ElkePahl) [reply src/MCRun.jl:257]: that's fine

File: src/MCRun.jl
Comment (ElkePahl) [reply src/MCRun.jl:278]: why? @inbounds could make it faster. Have you checked?

File: src/MCRun.jl
Comment (ElkePahl) [reply src/MCRun.jl:609]: That's good and we might also pull it oout further as discussed

File: src/MCRun.jl
Comment (ElkePahl) [reply src/MCRun.jl:707]: good. Do we want more specific names?

File: src/MCRun.jl
Comment (ElkePahl) [reply src/MCRun.jl:713]: That's good

File: src/MCRun.jl
Comment (ElkePahl) [inline src/MCRun.jl:808]: not so sure yet about if we really need this extra function?

File: src/MCRun.jl
Comment (ElkePahl) [reply src/MCRun.jl:924]: ok

File: src/MCRun.jl
Comment (ElkePahl) [reply src/MCRun.jl:933]: good

File: src/MCRun.jl
Comment (ElkePahl) [reply src/MCRun.jl:860]: What's left to be done?

File: src/MCRun.jl
Comment (ElkePahl) [inline src/MCRun.jl:949]: again why did you remove @inbounds?

**Review by ElkePahl — COMMENTED** [review]
Great job, Gray! I am mainly fine - but please have a look at my comments and decide if you want to change some of it before approval. If you don't do it immediately, please write it on a to-do-list and we can discuss further in the meeting on the clean-up.

File: src/MCRun.jl
Comment (grayseff) [reply src/MCRun.jl:77]: reinserted

File: src/MCRun.jl
Comment (grayseff) [reply src/MCRun.jl:707]: Presently no, eventually this will become a checkpoint function

File: src/MCRun.jl
Comment (grayseff) [reply src/MCRun.jl:808]: Moving multi-operation for-loops to functions is slightly cleaner. It also makes (for example) nested MC easier to implement.

File: src/MCRun.jl
Comment (grayseff) [reply src/MCRun.jl:860]: RDF and histograms haven't been checked on restart

File: src/MCRun.jl
Comment (ElkePahl) [reply src/MCRun.jl:139]: Could you please re-re-name here as well?

File: src/MCRun.jl
Comment (ElkePahl) [reply src/MCRun.jl:240]: There seemed to be a typo in line 186: ma**c**_acc instead of ma**x**_acc

again re-name, perhaps delta_en_move

File: src/MCRun.jl
Comment (ElkePahl) [reply src/MCRun.jl:816]: Yes, in principle I agree. But here I would rather have call of MC states and sampling extra so that we can use same functions for equilibration. But this is something we can discuss further before the clean up

File: src/MCRun.jl
Comment (ElkePahl) [reply src/MCRun.jl:868]: ok - keep this on your to-do-list :)

**Review by ElkePahl — COMMENTED** [review]
almost there - a very few comments left ...

File: src/MCRun.jl
Comment (grayseff) [reply src/MCRun.jl:240]: Done on both counts

**Review by ElkePahl — APPROVED** [review]
all fine now

**Pull request merged**
